src/MessageRouter.py

```diff
@@ -0,0 +1,47 @@
+import time
+import re
+
+
+class MessageRouter:
+    """
+    Responsible for invoking a message command handler if a message contains
+    a command and the rate limit is not exceeded.
+    """
+
+    def __init__(self,
+                 command_handlers={},
+                 default_rate_limit_in_seconds=60,
+                 per_command_rate_limit_in_seconds={}):
+        self._default_rate_limit = default_rate_limit_in_seconds
+        self._command_handlers = command_handlers
+        self._per_command_rate_limit = per_command_rate_limit_in_seconds
+
+        self._time_of_last_command_usage = {}
+
+    def handle_message(self, message):
```

Comment:
> with this overall structure it isn't possible to have a command handler send a private response to whoever wrote the message

The command handlers do not send anything. They get a string and return what the response for such a string would be. This is not bound to the IRC context, and can be used for whatever. You could create a CLI tool that uses these things. And you can definitely have the IRC context use it in different manners, such as having the result send to different channels. The responsibility of how the command handlers are used lies with their user, not with them, else you'd have a huge SRP violation.